onnxscript/evaluator.py

```diff
@@ -144,3 +204,99 @@ def call_ort(schema, args, kwargs, implicit_args=None):
     # Map ORT output values to the onnxscript representation-type.
     cast_result = [ort_to_os_value(x) for x in result]
     return cast_result[0] if len(cast_result) == 1 else cast_result
+
+
+def id(schema):
+    return schema.name, schema.domain, schema.since_version
+
+
+class ORTEvaluator(Evaluator):
+    """Evaluates ONNX ops using ONNX Runtime."""
+
+    def _eval(self, schema, inputs, attributes, closure):
+        return call_ort(schema, inputs, attributes, closure)
+
+
+ort_evaluator = ORTEvaluator()
+
+
+class ORTMixedEvaluator(ORTEvaluator):
+    """Evaluates ONNX ops using ONNX Runtime, unless an overriding python implementation
+    is registered. This is useful for higher-order ops such as Scan and SequenceMap,
+    allowing for python-based debugging.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._python_ops = {}
+
+    def use_graph_attribute(self, schema):
+        return id(schema) not in self._python_ops
+
+    def _eval(self, schema, inputs, attributes, closure):
+        if id(schema) in self._python_ops:
+            return self._python_ops[id(schema)](inputs, attributes)
+        else:
+            return super()._eval(schema, inputs, attributes, closure)
+
+    def register(self, opset: Optional[values.Opset] = None):
+        opset = opset or onnx_opset.default_opset
+
+        def decorator(function):
+            schema = opset[function.__name__]
+            self._python_ops[id(schema)] = function
+            return function
+
+        return decorator
+
+
+ort_mixed_evaluator = ORTMixedEvaluator()
+
+
+@ort_mixed_evaluator.register()
+def SequenceMap(inputs, attributes):
+    """Evaluates a SequenceMap op."""
+    fun = attributes["body"]
+
+    def get_input_of(input_index, iter_num):
+        input = inputs[input_index]
+        if isinstance(input, list):
+            return input[iter_num]
+        return input
+
+    def get_input(iter_num):
+        return [get_input_of(input_index, iter_num) for input_index in range(len(inputs))]
+
+    return [fun(*(get_input(i))) for i in range(len(inputs[0]))]
+
+
+# Used to control the default evaluator instance. A simple approach for now.
+
+default_ = ort_evaluator
```

Comment:
This looks like it can be private: `_default_evaluator`

Reply:
Done